WordPress/gutenberg
# [RNMobile] Add Featured Banner to Image Block (Android Only)

Paritial fix for: wordpress-mobile/gutenberg-mobile#1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3371
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/14451

## Description

This PR introduces a **Featured** banner, which overlays a post's featured image when it's added to an image block. This will make it easier for users to identify which of the images in their posts, if any, are set as featured directly from the editor.

_Note: This PR introduces the **Featured** banner on Android devices only, support for iOS will follow in future iterations._

## How has this been tested?

### Test Case 1: General Flow

To start with, an image block should be updated to reflect whether its image is featured or not when set via the general Post Settings flow:

* Tap the three icons to the upper right of the post editor to access **Post Settings** and set a featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner displays overlays the correct image block.
* Return to **Post Settings**. Tap the existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the image block, as expected.

### Test Case 2: "Editor Loads" Flow

Next, we should verify that a **Featured** banner appears when a post with a featured image is first loaded. To test this, follow these steps:

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it.

### Test Case 3: "Replace in Image Block" Flow

Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the **Featured** banner updates as expected depending on the image that's chosen for an image block:

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/114570411-b7907c80-9c6d-11eb-9d3c-8672b0771d0b.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/114570332-a5164300-9c6d-11eb-88f7-fe751259087e.png" width="100%"> |

## Types of changes

This PR introduces a new feature (a non-breaking change that adds functionality). Specifically, it adds a **Featured** banner that overlays an image block containing a post's featured image. The code can be split into two overarching categories:

### Badge Component

* A [new Badge component](https://github.com/WordPress/gutenberg/pull/30806/files#diff-f3d1866bb6de6b07c40f6cc1e97d08ef289dc2b4874deec8c414541e0dd5844e) has been created, which adds a "badge" with some text to its child component's upper left. 
* In this case, this new component is used to add a **Featured** badge to an image block. 
* The component was created with re-usability in mind and [the README](https://github.com/WordPress/gutenberg/pull/30806/files#diff-2500f619901448109c703529708b09a157df9072a96e99c19d1d217d54e9aa24) describes how it may be used in other areas of the codebase.

### Logic for Fetching Featured Image ID

* To fetch the ID of a post's initial featured image, [a prop is passed over the bridge](https://github.com/WordPress/gutenberg/pull/30806/files#diff-906aa52029680055db7dc413437ac227ebeaef57c0da27347a019fc53506099a) and set in [the post object](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94R146).
* An [event subscription](https://github.com/WordPress/gutenberg/pull/30806/files#diff-bf4c994cef3006848617e4cd8a8ad4bfbc79ef6feda4d58152369ec92019d51e) is used to track a change to a post's featured image from the Android app. The `editEntityRecord` selector is then used in combination with listeners to [update the value stored for a post's featured image ID](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94). 
* The above enables a `featuredImageId` to be passed to and [used in the image component's edit.native.js file](https://github.com/WordPress/gutenberg/pull/30806/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1). In order to determine whether or not to display the badge component, the `featuredImageId` is compared to the ID of the image in the block.

## Checklist:
